### PR TITLE
Add Filter&Alarm !Exists For Network Gateways

### LINKED ISSUE #1173: Add Filter And Alarm Not Exists For Changes To Network Gateways query for CloudFormation
### Platform
CloudFormation

### Provider
AWS

### Description
A log metric filter and alarm should exist for Changes to Network Gateways

### PR DESCRIPTION
**Pull request by fabioGoncalvesCx**
Added new query for AWS CloudFormation.

Provide a CloudFormation resource and a resource to ensure that Network Gateways changes are catched by a log metric filter and an alarm
Closes #1173

### CODE REVIEW

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/metadata.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "Filter_And_Alarm_Exist_For_S3_Changes_To_Network_Gateways",
+  "queryName": "Filter And Alarm Not Exists For Changes To Network Gateways",
+  "severity": "MEDIUM",
+  "category": "Monitoring",
+  "descriptionText": "A log metric filter and alarm should exist for Changes to Network Gateways",
+  "descriptionUrl": "https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudwatch-alarms-for-cloudtrail.html"
+}
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/metadata.json:8]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
+    resourceAlarm.Namespace != metricFilter
+}
+
+checkTopic(resourceAlarm, nameSNS) = true {
+    resourceAlarm.Ref != nameSNS
+}
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego:74]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/negative.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  MetricFilter:
+    Type: 'AWS::Logs::MetricFilter'
+    Properties:
+      LogGroupName: ''
+      FilterPattern: >-
+        { ($.eventName = CreateCustomerGateway) || ($.eventName = DeleteCustomerGateway) || ($.eventName = AttachInternetGateway) || ($.eventName = CreateInternetGateway) || ($.eventName = DeleteInternetGateway) || ($.eventName = DetachInternetGateway) }
+      MetricTransformations:
+        - MetricValue: '1'
+          MetricNamespace: CloudTrailMetrics
+          MetricName: IAMPolicyEventCount
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/negative.yaml:38]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/positive.yaml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  MetricFilter3:
+    Type: 'AWS::Logs::MetricFilter'
+    Properties:
+      LogGroupName: ''
+      FilterPattern: >-
+        { ($.eventName = CreateCustomerGateway) || ($.eventName = DeleteCustomerGateway) || ($.eventName = AttachInternetGateway) || ($.eventName = CreateInternetGateway) || ($.eventName = DeleteInternetGateway) || ($.eventName = DetachInternetGateway) }
+      MetricTransformations:
+        - MetricValue: '1'
+          MetricNamespace: CloudTrailMetrics
+          MetricName: IAMPolicyEventCount
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/positive.yaml:109]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.FilterPattern' should have the right specification')", [name]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
+    resourceAlarm.Namespace != metricFilter
+}
+
+checkTopic(resourceAlarm, nameSNS) = true {
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego:72]: This function is only being used once and should not be declared, replace it for a direct comparison

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	     sprintf("Resources.%s.Properties.FilterPattern", [name]),
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.FilterPattern' should have the right specification')", [name]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/query.rego:68]: This function is only being used once and should not be declared, replace it for a direct comparison

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/negative.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        attaching/detaching policies from IAM users, roles or groups.
+      MetricName: IAMPolicyEventCount
+      Namespace: CloudTrailMetrics
+      Statistic: Sum
+      Period: 300
+      EvaluationPeriods: 1
+      Threshold: 1
+      ComparisonOperator: GreaterThanOrEqualToThreshold
+      AlarmActions:
+        - Ref: SnsTopic
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/test/negative.yaml:27]: Either remove the reference (intrinsic function `Ref`) to Resource that doesn't exists or add it as a [Parameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html) to the header of the template

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/metadata.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "Filter_And_Alarm_Exist_For_S3_Changes_To_Network_Gateways",
+  "queryName": "Filter And Alarm Not Exists For Changes To Network Gateways",
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_for_changes_to_network_gateways/metadata.json:3]: Change **directory name** and **id** to: `network_gateway_changes_without_alarms`
Change **query name** to: `Network Gateway Changes Without Alarms`

**Review by rogeriopeixotocx — CHANGES_REQUESTED** [review]
Requesting changes

**Comment by rogeriopeixotocx** [comment]
Fix the unit tests @fabioGoncalvesCx

**Comment by markmishaevcx** [comment]
Scan submitted to Checkmarx

**Comment by markmishaevcx** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 4 vulnerabilities<br />High&nbsp;**0 High**<br />Medium&nbsp;**3 Medium**<br />Low&nbsp;**1 Low**<br />Info&nbsp;**0 Info**<br />
### Violation Summary
Medium&nbsp;**3** **Medium**<br />
[View more details on Checkmarx UI](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049055&projectid=726)
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[30](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_changes_to_network_gateways/cmd/builder/main.go#L30) |Medium|Path_Traversal|cmd/builder/main.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049055&projectid=726&pathid=3)
[80](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_changes_to_network_gateways/test/main_test.go#L80) |Medium|Denial_Of_Service_Resource_Exhaustion|test/main_test.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049055&projectid=726&pathid=2)
[52](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_changes_to_network_gateways/pkg/kics/service.go#L52) |Medium|Denial_Of_Service_Resource_Exhaustion|pkg/kics/service.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049055&projectid=726&pathid=1)

**Review by rogeriopeixotocx — APPROVED** [review]
LGTM

**Pull request merged**